src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -1011,25 +1010,41 @@ private void BlockDownloaded(uint256 blockHash, Block block, int peerId)
                     throw new InvalidOperationException("Unsolicited block");
                 }
 
-                if (block != null)
+                chainedHeader = this.chainedHeaderTree.GetChainedHeader(blockHash);
```

Comment:
I think `BlockDownloaded` is not the right place for this code. 

`BlockDownloaded` is a low level callback from BP. Actual consumer might be another component, not CHT. 

For example we are a pruned node and blocks 0-100k are pruned. Wallet is syncing from 50k. Blocks that are requested are not marked as requested by CHT because they are already part of consensus. 

So this code should be only for blocks that were requested by CHT itself.

Reply:
Ok but this code was there before this change, the only reason we get the chained header earlier is so that we can construct ChainedHeaderBlock later when we pass it to ProcessDownloadedBlock.

Reply:
If you look at how it was done before we only get ChainedHeader if Block is not null... So if we do it like that we can't construct ChainedHeaderBlock.

Reply:
I don't agree @noescape00, this is just to fix the race condition so it should be in the lower level code.

Reply:
>I don't agree @noescape00, this is just to fix the race condition so it should be in the lower level code.

So let's say we are in this scenario with pruned node. CT is at 200k, earlist block we have is 100k, we need some block below 100k. Wallet asks for it. It's `BlockDataAvailabilityState` is still `HeaderOnly`, not `BlockRequired` as in case when CHT asked for it. 

So the BP fails for some reason and we go into `BlockDownloaded` with block being `null`. On 1029 we won't go into that if and `reassignDownload` will be false. Now we will go to 1075 (and here you are suggesting to throw in another review message, but it's a valid scenario according to current logic).

Reply:
Ok I think we are not dicussing anymore the fact tha the code for 
```
chainedHeader = this.chainedHeaderTree.GetChainedHeader(blockHash); 
```
Is outside the if `block != null` condition right?

Logically speaking this change is only to bypass the potential race condition, if the header is in the state of `BlockRequired` we have to either succeed or remove the header this is the logic I understand to be correct, but you claim that may not be the case, in that case if block download failed we have to set the state of the header back to HeaderOnly no?